include/ginkgo/core/base/range_accessors.hpp

```diff
@@ -205,6 +631,289 @@ class row_major {
 };
 
 
+namespace detail_colmajor {
+
+
+/**
+ * This helper runs from first to last dimension in order to compute the index.
+ * The index is computed like this:
+ * indices: x1, x2, x3, ..., xn
+ * compute(stride, x1, x2, x3, ..., x(n-1), xn) ->
+ *  x1 * stride[0] + x2 * stride[1] + ...
+ *    + x(n-2) * stride[n-3] + x(n-1) + xn * stride[n-2]
+ * Note that swap of the last two strides, making this 'column major'.
+ */
+template <typename ValueType, size_type total_dim, size_type current_iter = 1>
+struct index_helper_s {
+    static_assert(total_dim >= 1, "Dimensionality must be >= 1");
+    static_assert(current_iter <= total_dim, "Iteration must be <= total_dim!");
+
+    static constexpr size_type dim_idx{current_iter - 1};
+
+    template <typename FirstType, typename... Indices>
+    static constexpr GKO_ATTRIBUTES ValueType
+    compute(const std::array<ValueType, total_dim> &size,
+            const std::array<ValueType, (total_dim > 0 ? total_dim - 1 : 0)>
+                &stride,
+            FirstType first, Indices &&... idxs)
+    {
+        if (current_iter == total_dim - 1) {
+            return first +
+                   index_helper_s<ValueType, total_dim, current_iter + 1>::
+                       compute(size, stride, std::forward<Indices>(idxs)...);
+        }
+
+        return GKO_ASSERT(first < size[dim_idx]),
+               first * stride[dim_idx] +
+                   index_helper_s<ValueType, total_dim, current_iter + 1>::
+                       compute(size, stride, std::forward<Indices>(idxs)...);
+    }
+};
+
+template <typename ValueType, size_type total_dim>
+struct index_helper_s<ValueType, total_dim, total_dim> {
+    static_assert(total_dim >= 2, "Dimensionality must be >= 2");
+
+    static constexpr size_type dim_idx{total_dim - 1};
+
+    template <typename FirstType, typename... Indices>
+    static constexpr GKO_ATTRIBUTES ValueType
+    compute(const std::array<ValueType, total_dim> &size,
+            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>
+                &stride,
+            FirstType first, Indices &&...)
+    {
+        return GKO_ASSERT(first < size[total_dim - 1]),
+               first * stride[dim_idx - 1];
+    }
+};
+
+/**
+ * Computes the storage index for the given indices with respect to the given
+ * stride array
+ */
+template <typename ValueType, size_type total_dim, typename... Indices>
+constexpr GKO_ATTRIBUTES ValueType compute_index(
```

Comment:
Maybe rename the previous one to `compute_row_major_index` and this one to `compute_col_major_index`, so both can be in the namespace `detail`, and it is clear what each is for.

Reply:
Well.. there's still stuff like `default_stride_array_impl` below which will become too long if I do that. Especially if we rename the accessor as block_col_major. Maybe I can just add a line of documentation for `detail_colmajor` to make things clear? (I guess this namespace will then be renamed `detail_block_col_major`.)

Reply:
I think a comment for `compute_storage_index` might be needed, since there is no information if it is column- or row-major.
For this one, you are right, the namespace should be enough (in my mind, I was fusing both namespaces and would have a clash there).